megazords/ios/RustLog/RustLog.swift

```diff
@@ -0,0 +1,254 @@
+    ///
+    /// - RustLogError.UnexpectedError: If the rust code panics. This shouldn't happen,
+    ///   but if it does, we would appreciate reports from telemetry or similar
+    public static func enable(_ callback: @escaping LogCallback) throws {
+        try queue.sync {
+            try state.enable(callback)
+        }
+    }
+
+    /// Set the level filter (the maximum log level) of the logger.
+    ///
+    /// Throws:
+    /// - RustLogError.UnexpectedError: If the rust code panics. This shouldn't happen,
+    ///   but if it does, we would appreciate reports from telemetry or similar
+    public static func setLevelFilter(filter: LogLevelFilter) throws {
+        // Note: Doesn't need to synchronize.
+        try rustCall { error in
+            rc_log_adapter_set_max_level(filter.toInt(), error)
+        }
+    }
+
+    /// Disable the previously set logger. This also sets the level filter to Off.
+    ///
+    /// Does nothing if the logger is disabled
+    public static func disable() {
+        queue.sync {
+            state.disable()
+        }
+    }
+
+    /// Enable the logger if possible.
+    ///
+    /// Returns false in the cases where `enable` would throw, true otherwise.
+    ///
+    /// If it would throw due to a panic, it also writes some information about
+    /// the panic to the provided callback
+    public static func tryEnable(_ callback: @escaping LogCallback) -> Bool {
+        return queue.sync {
+            state.tryEnable(callback)
+        }
+    }
+
+    /// Log a test message at `Info` severity.
+    public static func logTestMessage(message: String) {
+        rc_log_adapter_test__log_msg(message)
+    }
+}
+
+/// The type of errors reported by RustLog. These either indicate bugs
+/// in our logging code (as in `UnexpectedError`), or usage errors
+/// (as in `AlreadyEnabled`)
+public enum RustLogError: Error {
+    /// This generally means a panic occurred, or something went very wrong.
+    /// We would appreciate bug reports about when these appear in the wild, if they do.
+    case UnexpectedError(message: String)
+
+    /// Error indicating that the log adapter cannot be enabled
+    /// because it is already enabled.
+    ///
+    /// This is a usage error, either `disable` it first, or
+    /// use `RustlLog.tryEnable`
+    case AlreadyEnabled
+}
+
+@discardableResult
+fileprivate func rustCall<T>(_ callback: (UnsafeMutablePointer<RcLogError>) throws -> T) throws -> T {
+    var err = RcLogError(code: 0, message: nil)
+    let result = try callback(&err)
+    if err.code != 0 {
+        let message = err.message!
+        defer { rc_log_adapter_destroy_string(message) }
+        throw RustLogError.UnexpectedError(message: String(cString: message))
+    }
+    return result
+}
+
+// This is the function actually passed to Rust.
+fileprivate func logCallbackFunc(level: Int32, optTagP: UnsafePointer<CChar>?, msgP: UnsafePointer<CChar>) {
+    if let callback = RustLog.state.callback {
```

Comment:
nit: swifty style of `guard let callback = ... else { return }` preferred to reduce closing brace cascade at function end